src/QsCompiler/DataStructures/DataTypes.fs

```diff
@@ -53,61 +53,164 @@ type QsNullable<'T> = // to avoid having to include the F# core in the C# part o
+            else relative.Column
+        Position (line, column)
+
+    /// <summary>
+    /// Translates the first position backwards by the amount of the second position. If the resulting position is on
+    /// the same line, the column numbers are subtracted; otherwise, the first position's column number is used.
+    /// </summary>
+    /// <exception cref="ArgumentOutOfRangeException">
+    /// Thrown if the resulting position has a negative line or column number.
+    /// </exception>
+    static member (-) (position : Position, offset : Position) =
+        let line = position.Line - offset.Line
+        let column =
+            if position.Line = offset.Line
+            then position.Column - offset.Column
+            else position.Column
+        Position (line, column)
+
+    /// Returns true if the positions have the same line and column numbers.
+    static member op_Equality (a : Position, b : Position) = a = b
+
+    /// Returns true if the second position occurs before the first position.
+    static member op_LessThan (a : Position, b) = (a :> Position IComparable).CompareTo b < 0
+
+    /// Returns true if the second position occurs on or before the first position.
+    static member op_LessThanOrEqual (a : Position, b) = (a :> Position IComparable).CompareTo b <= 0
+
+    /// Returns true if the second position occurs after the first position. 
+    static member op_GreaterThan (a : Position, b) = (a :> Position IComparable).CompareTo b > 0
+
+    /// Returns true if the second position occurs on or after the first position.
+    static member op_GreaterThanOrEqual (a : Position, b) = (a :> Position IComparable).CompareTo b >= 0
+
+    /// <summary>
+    /// Creates a position.
+    /// </summary>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown if the line or column is negative.</exception>
+    static member Create line column =
+        if line < 0 || column < 0
+        then ArgumentOutOfRangeException "Line and column cannot be negative." |> raise
+        Position (line, column)
+
+    /// The position at line zero and column zero.
+    static member Zero = Position (0, 0)
+
+/// A range between two positions in a text document.
+type Range = private Range of Position * Position with
+
+    /// The start of the range.
+    member this.Start = match this with Range (start, _) -> start
+
+    /// The end of the range.
+    member this.End = match this with Range (_, end') -> end'
+
+    /// Returns true if the range contains the given position, excluding the end position.
+    member this.Contains position = this.Start <= position && position < this.End
+
+    /// Returns true if the range contains the given position, including the end position.
+    member this.ContainsEnd position = this.Start <= position && position <= this.End
+
+    /// Translates the line numbers of the start and end positions by the given offset.
+    member this.TranslateLines offset =
```

Comment:
Even though the naming here is clearer, I would still go with something like WithLineNrOffset.

Reply:
See above, I copied this to the first `TranslateLines` comment.